nullaway/src/main/java/com/uber/nullaway/GenericsCheck.java

```diff
@@ -0,0 +1,145 @@
+    if (!config.isJSpecifyMode()) {
+      throw new IllegalStateException("should only check type instantiations in JSpecify mode");
+    }
+    List<? extends Tree> typeArguments = tree.getTypeArguments();
+    HashSet<Integer> nullableTypeArguments = new HashSet<Integer>();
+    JCTree.JCTypeApply baseTypeTree = (JCTree.JCTypeApply) tree;
+    Type t = baseTypeTree.type;
+    Type.ClassType classTyp = (Type.ClassType) t;
+    if (classTyp.tsym.getMetadata() != null) {
+      List<Attribute.TypeCompound> baseTypeAttributes =
+          classTyp.tsym.getMetadata().getTypeAttributes();
+      for (int i = 0; i < baseTypeAttributes.size(); i++) {
+        nullableTypeArguments.add(baseTypeAttributes.get(i).position.parameter_index);
+      }
+    }
+    for (int i = 0; i < typeArguments.size(); i++) {
+      boolean hasNullableAnnotation = false;
+      if (typeArguments.get(i).getClass().equals(JCTree.JCTypeApply.class)) {
+        ParameterizedTypeTree parameterizedTypeTreeForTypeArgument =
+            (ParameterizedTypeTree) typeArguments.get(i);
+        Type argumentType = ASTHelpers.getType(parameterizedTypeTreeForTypeArgument);
+        if (argumentType != null
+            && argumentType.getTypeArguments() != null
+            && argumentType.getTypeArguments().length() > 0) { // Nested generics
+          checkInstantiationForParameterizedTypedTree(
+              parameterizedTypeTreeForTypeArgument, state, currentState, config);
+        }
+      }
+      if (typeArguments.get(i).getClass().equals(JCTree.JCAnnotatedType.class)) {
+        JCTree.JCAnnotatedType annotatedType = (JCTree.JCAnnotatedType) typeArguments.get(i);
+        for (JCTree.JCAnnotation annotation : annotatedType.getAnnotations()) {
+          Attribute.Compound attribute = annotation.attribute;
+          hasNullableAnnotation =
+              hasNullableAnnotation
+                  || attribute.toString().equals("@org.jspecify.nullness.Nullable");
+          if (hasNullableAnnotation) {
+            break;
+          }
+        }
+      }
+      if (hasNullableAnnotation) {
+        if (!nullableTypeArguments.contains(i)) {
+          GenericsCheck.invalidInstantiationError(
+              typeArguments.get(i), state, currentState, config);
+        }
+      }
+    }
+  }
+
+  private static void invalidInstantiationError(
+      Tree tree, VisitorState state, NullAway currentState, Config config) {
+    ErrorBuilder errorBuilder = new ErrorBuilder(config, "", ImmutableSet.of());
+    ErrorMessage errorMessage =
+        new ErrorMessage(
+            ErrorMessage.MessageTypes.TYPE_PARAMETER_CANNOT_BE_NULLABLE,
+            "Generic type parameter cannot be @Nullable");
+    state.reportMatch(
+        errorBuilder.createErrorDescription(
+            errorMessage, currentState.buildDescription(tree), state, null));
+  }
+
+  public static void checkNestedParameterInstantiation(
```

Comment:
Rather than having this as a separate method, probably the recursion should be integrated into `checkInstantiatedType`